Assets/_BForBoss/_Character/Scripts/PlayerWallRunBehaviour.cs

```diff
@@ -0,0 +1,232 @@
+using ECM2.Characters;
+using ECM2.Components;
+using ECM2.Common;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+using System;
+using UnityEngine.InputSystem;
+
+namespace BForBoss
+{
+    public class PlayerWallRunBehaviour : MonoBehaviour
+    {
+        [SerializeField] 
+        private float _speedMultiplier = 1f;
+        [SerializeField, Tooltip("Don't allow a wall run if the player is too close to the ground")] 
+        private float _minHeight = 1f;
+        [SerializeField]
+        private float _wallMaxDistance;
+        [SerializeField]
+        private float _wallGravityDownForce = 0f;
+        [SerializeField, Range(0f, 3f), Tooltip("Only allow for a wall run if jump is longer than this")]
+        private float _minJumpDuration = 0.3f;
+
+        [SerializeField]
+        private float _maxCameraAngleRoll;
+        [SerializeField]
+        private float _cameraRotateDuration;
+
+        private readonly Vector3[] directions = new Vector3[]
+        {
+            Vector3.right, 
+            Vector3.right + Vector3.forward,
+            Vector3.forward,
+            Vector3.left + Vector3.forward,
+            Vector3.left
+        };
+
+        private bool _isWallRunning = false;
+        private Character _baseCharacter = null;
+        private FirstPersonCharacter _fpsCharacter = null;
+        private RaycastHit[] _hits;
+        private LayerMask _mask;
+        private Vector3 _lastWallRunPosition;
+        private Vector3 _lastWallRunNormal;
+        private float _baseMaxSpeed;
+        private Func<Vector2> _movementInput;
+
+        private float _currentJumpDuration = 0f;
+        private float _timeSinceWallAttach = 0f;
+        private float _timeSinceWallDetach = 0f;
+
+        private Transform ChildTransform
+        {
+            get
+            {
+                return _fpsCharacter != null ? _fpsCharacter.rootPivot : transform;
+            }
+        }
+
+        public void Initialize(Character baseCharacter, Func<Vector2> getMovementInput)
+        {
+            _baseCharacter = baseCharacter;
+            _fpsCharacter = baseCharacter as FirstPersonCharacter;
+            _movementInput = getMovementInput;
+            _mask = LayerMask.GetMask("ParkourWall");
+        }
+
+        public void Falling(Vector3 _)
+        {
+            if (!CanWallRun()) return;
+            _hits = new RaycastHit[directions.Length];
+            for (int i = 0; i < directions.Length; i++)
+            {
+                Vector3 dir = ChildTransform.TransformDirection(directions[i]); //convert directions to relative to player
```

Comment:
can you remove the comment and make this variable more verbose? 
e.g. `playerFacingDirection`